docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,318 @@
+Improving redirects
+===================
+
+Redirects are a core feature of Read the Docs,
+they allow users to keep old URLs working when they rename or move a page.
+
+The current implementation lacks some features and has some undefined/undocumented behaviors.
+
+.. contents::
+   :local:
+   :depth: 3
+
+Goals
+-----
+
+- Improve the user experience when creating redirects.
+- Improve the current implementation without big breaking changes.
+
+No goals
+--------
+
+- Replicate every feature of other services without
+  having a clear use case for them.
+- Improve the performance of redirects.
+  This can be discussed in an issue or pull request.
```

Comment:
I do think performance is a consideration in the design here, if there are ways to make it faster/cheaper, but not a core priority, I'd agree.